### PR TITLE
Fail closed on coalesced turns the record cannot attribute

### LINKED ISSUE #1670: Replay guard fails open for pre-metadata coalesced ledger records
Follow-up to #1667, which fixed mixed-requester replay drops for records that carry per-source metadata but leaves one corner open.

## Problem

`src/mindroom/text_ingress_dispatch.py:318`:

```python
may_be_superseded = prepared.dispatch.envelope.origin.may_be_superseded_by_newer_requester_turn and all(
    metadata.sender == requester_user_id
    for metadata in (prepared.handled_turn.source_event_metadata or {}).values()
)
```

When `source_event_metadata is None`, `all([])` is `True`, so `may_be_superseded` keeps its original value and a coalesced batch can still be settled without a response — the exact failure #1667 set out to prevent.

Confirmed on merged `main` (`8f587f3d3`):

```
is_coalesced=True  source_event_metadata=None
#1667 all(...) over absent metadata -> True
  => may_be_superseded survives: YES (fail-open)
```

## Exposure

Narrow. Requires all three:

1. a ledger record with `len(source_event_ids) > 1` and `source_event_metadata is None`
2. that record still unresolved at dispatch
3. its sources spanning multiple requesters

Current code cannot produce (1): `turn_controller.py:880` and `turn_controller.py:1877` set `source_event_metadata` whenever `len(source_event_ids) > 1`. The only source is records persisted **before** `source_event_metadata` existed (#1342, #1464).

Those records still decode. `handled_turns.py:410` rejects payloads whose `schema_version` differs from the codec version, but `_TURN_RECORD_SCHEMA_VERSION` is still `1` and the field was added under that same version without a bump, so a pre-metadata coalesced record round-trips with the field absent.

Not a regression from #1667 — that PR strictly improved the behaviour and simply did not cover this shape.

## Inconsistency

`TurnRecord.requester_id_for_source` (landing with #1641) answers the same question and deliberately fails **closed** for coalesced turns without metadata, returning `None`. Two files currently give opposite answers to "who requested this source?".

## Suggested fix

Once #1641 lands, collapse the guard onto the shared helper so the rule lives in one place and fails closed on the legacy shape:

```python
may_be_superseded = prepared.dispatch.envelope.origin.may_be_superseded_by_newer_requester_turn and all(
    prepared.handled_turn.requester_id_for_source(source_event_id) == requester_user_id
    for source_event_id in prepared.handled_turn.source_event_ids
)
```

Behaviour-identical for every record current code writes; fail-closed for the legacy shape.

## Test to add

A coalesced `TurnRecord` with multiple `source_event_ids`, `source_event_metadata=None`, and mixed requesters must not be settled without a response.

### PR DESCRIPTION
**Pull request by basnijholt**
Follow-up to #1667. Closes #1670.

## Why

Whole-turn suppression settles every physical source in a coalesced batch, so it is only safe when the turn provably belongs to one requester.

#1667 proved that from `source_event_metadata`, but read the check off the map's values:

```python
all(metadata.sender == requester_user_id for metadata in (handled_turn.source_event_metadata or {}).values())
```

That is vacuously true for any source the map does not cover, in two distinct shapes:

1. **Metadata absent entirely.** `all([])` is `True`, so `may_be_superseded` survived and a coalesced batch could be settled without a response — the failure #1667 set out to prevent.
2. **Metadata covering only some sources.** `.values()` never looks at the sources it is missing, so a map holding just `$bob` reads as "every source is bob's" and settles `$alice` too.

Confirmed on merged `main` (`8f587f3d3`):

```
is_coalesced=True  source_event_metadata=None
#1667 all(...) over absent metadata -> True
  => may_be_superseded survives: YES (fail-open)
```

## Exposure

Narrow, and not a regression from #1667 — that PR strictly improved the behaviour and simply did not cover these shapes.

**Absent metadata.** Current code cannot write it: `turn_controller.py:880` and `turn_controller.py:1877` set `source_event_metadata` whenever `len(source_event_ids) > 1`. The reachable source is records persisted **before** the field existed (#1342, #1464). They still decode: `handled_turns.py:809` rejects payloads whose `schema_version` differs from the codec version, but `_TURN_RECORD_SCHEMA_VERSION` is still `1` and the field was added under that same version without a bump, so a pre-metadata coalesced record round-trips with the field absent.

**Partial metadata.** `_immutable_source_event_metadata` keeps only entries it can normalize, so a map that lost one source's entry survives as a valid, partially populated map. `_backfill_missing_turn_facts` can also pair a ledger record's identity with a recovery record's metadata when the ledger's is `None`, without requiring the two to cover the same sources.

## What changed

`edit_regenerator.py` already resolved "who owns this source" inline, with exactly the no-metadata fallback, per-source lookup, and fail-closed-on-miss the replay guard needs — `test_partial_coalesced_metadata_rejects_anchor_sender_editing_sibling` has pinned that rule for the edit path for a while. Rather than write a second implementation, the rule moves onto `TurnRecord` beside `prompt_source_event_id`:

```python
def requester_id_for_source(self, event_id: str) -> str | None:
    if self.source_event_metadata is None:
        return self.requester_id if not self.is_coalesced else None
    metadata = self.source_event_metadata.get(self.prompt_source_event_id(event_id))
    return metadata.sender if metadata is not None else None
```

Both callers use it. The replay guard asks it of every replayable source:

```python
return all(
    handled_turn.requester_id_for_source(source_event_id) == requester_user_id
    for source_event_id in handled_turn.replay_source_event_ids
)
```

`replay_source_event_ids` rather than `source_event_ids`: normalization prunes a redacted source's metadata entry by design, so demanding an entry for every source id would make any coalesced turn containing a redaction permanently unsuppressible, even though a redacted source owes no reply.

Behaviour is unchanged for every record current code writes — `_batch_source_event_metadata` keys an entry off every pending event in the batch, so live coalesced records are always fully covered.

## Note for #1641

#1641 adds an identical `TurnRecord.requester_id_for_source`. Whichever merges second drops its copy; the method body here matches #1641's semantics exactly, so the resolution is mechanical.

## Validation

#1667's ownership matrix now parameterizes the shape the record carries — `single | mixed | absent | partial | redacted` × `degraded`, ten cases — instead of a boolean, so the new shapes are covered on the degraded path too. Record construction lives in `_coalesced_ownership_record`, which pins the normalization invariants each case name claims (`absent` normalizes to `None`; `partial` and `redacted` normalize to a `$bob`-only map). The parameter list is derived from the `Literal` that types it, so the two cannot drift.

Mutation-tested, each mutation run against the full matrix:

| Mutation | Result |
| --- | --- |
| Restore the `.values()` check | `partial-False`, `partial-True` fail (`assert [] == [PreparedDisp..._budget=None)]`) |
| Iterate `source_event_ids` instead of `replay_source_event_ids` | `redacted-False`, `redacted-True` fail |
| Uncovered source falls back to the turn-level requester | `partial-False`, `partial-True` **and** `test_partial_coalesced_metadata_rejects_anchor_sender_editing_sibling` fail — both callers enforce the shared rule |
| Coalesced no-metadata records claim the turn-level requester | `absent-False`, `absent-True` fail |
| `return False` for the no-metadata branch | 8 existing backlog-replay tests fail across `test_live_message_coalescing.py` and `test_turn_controller_focused.py` |

- `tests/test_live_message_coalescing.py` plus every edit-path suite: pass.
- Exact-head full suite: `11874` passed, `54` skipped, zero failures or errors.
- Exact-head GitHub pytest, smoke, builds, static checks, and AI review checks are green.
- Tach dependencies/interfaces and every all-file pre-commit hook pass.
- The frozen real-Tuwunel replay passed `200` operations with `45` roots, `135` canonical replies, one restart, one outage, and zero coordinator timeouts, dispatch timeouts, or event-loop stalls.
- Three runtime generations attest exact clean MindRoom `83c1a62e5` and nio `e15f9e19`; final source validation and physical cleanup pass.
- Exact-head production source diff: `+56/-18`, net `+38`, across the ownership, replay, and cleanup paths.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved fail-closed handling for coalesced replay, especially when source ownership metadata is missing, partial, redacted, or involved in discovery/alias collisions.
  * Updated requester attribution checks to use per-source ownership resolution for edit regeneration.
  * Preserved explicit recovered `source_event_metadata` (including empty/unknown ownership) and improved redaction cleanup behavior to avoid tombstoning colliding aliases.

* **Tests**
  * Expanded regression tests for coalesced replay suppression, alias-collision edge cases, edit regeneration, redaction cleanup, codec persistence, and recovery scenarios.

## Follow-up: redaction cleanup after alias projection

Exact-head review found a second use for turn-level requester context. When a completed relay conflicts with an older coalesced turn and a redacted physical source shares the relay discovery alias, alias projection retains the physical source with an explicit empty metadata map. That map correctly keeps per-source replay ownership fail-closed, but clearing `requester_id` also removed the conversation identity needed to sanitize recorded scopes. The pending cleanup marker was then acknowledged without removing the redacted run, summary, and seen-event state.

`handled_turns.py` now preserves the turn-level requester through projection while retaining the explicit empty source map. `requester_id_for_source()` therefore still returns `None`, but `prepare_response_for_redactions()` has the context needed to sanitize before clearing the marker.

Regression evidence on `83c1a62e5`:

- The exact colliding relay/physical-source cleanup test reproduced RED before the source change and passes after it.
- Full `test_handled_turns.py`, `test_turn_store.py`, and `test_edit_regenerator.py` suites pass.
- All 22 backlog-replay tests pass.
- Changed-file pre-commit hooks pass, including ruff, format, `ty`, Vulture, Tach, and module privacy.

### CODE REVIEW

**Comment by basnijholt** [comment]
Review follow-up pushed in 9bc8b10b0.

**1. Partial per-source coverage was still fail-open.** `all(... for metadata in source_event_metadata.values())` only checked the entries the map happens to carry, and `_immutable_source_event_metadata` drops entries it cannot normalize and prunes redacted sources — so a coalesced record covering only `$bob` read as "every source is bob's" and suppressed `$alice` too. Same class as the absent-metadata bug, one shape over, and #1641's `requester_id_for_source` already fails closed here, so the collapse target was stricter than what this PR left behind.

Now iterates sources rather than values, over `replay_source_event_ids` so redacted sources (which own no reply) stay suppressible:

```python
return all(
    (metadata := source_event_metadata.get(source_event_id)) is not None and metadata.sender == requester_user_id
    for source_event_id in handled_turn.replay_source_event_ids
)
```

**2. Test duplication.** The standalone `test_backlog_replay_keeps_coalesced_turn_without_source_metadata` was a near-verbatim copy of #1667's already-parameterized matrix. Folded in as an `ownership` axis — `single | mixed | absent | partial | redacted` × `degraded` — so the legacy shape now also gets degraded-path coverage it previously lacked, and the record shapes are built once in `_coalesced_ownership_record` (which pins the normalization invariants: `absent` normalizes to `None`, `partial`/`redacted` normalize to `$bob`-only).

**3. Naming/docstrings.** `_turn_belongs_to_single_requester` -> `_turn_sources_all_from_requester`: it returns False for a turn that does belong to one requester who simply isn't this one. Test docstring said "empty metadata mapping" where `__post_init__` normalizes empty maps to `None`.

**Mutation evidence**
- Revert to the previously submitted `.values()` helper: `partial-False` and `partial-True` fail (`assert [] == [PreparedDisp..._budget=None)]`); all other params still pass.
- Iterate `source_event_ids` instead of `replay_source_event_ids`: `redacted-False` and `redacted-True` fail — pinning that a redacted source must not permanently block suppression.

**Validation**: full suite green locally, `tests/test_live_message_coalescing.py` 159 passed, all pre-commit hooks pass (ruff, format, ty, vulture, Tach, module privacy). Production diff is now +21/-3 in `src/mindroom/text_ingress_dispatch.py`.

File: src/mindroom/text_ingress_dispatch.py [hunk 1]
```diff
@@ -293,6 +293,20 @@ def _parsed_command_for_event(
...
+
+    Whole-turn suppression settles every source in a coalesced batch, so it is only safe when the
+    turn provably belongs to one requester. A coalesced record without per-source metadata cannot
+    prove that: records persisted before the metadata field existed still decode under the current
+    ledger schema version, so this fails closed rather than treating them as single-requester.
+    """
+    source_event_metadata = handled_turn.source_event_metadata
+    if source_event_metadata is None:
+        return not handled_turn.is_coalesced
+    return all(metadata.sender == requester_user_id for metadata in source_event_metadata.values())
```
[the remark this answers is not in the document]
Comment (basnijholt) [reply src/mindroom/text_ingress_dispatch.py]: Correct, and already fixed — 9bc8b10b0 landed before this comment posted, and f3a77b168 consolidated it.

One deviation from the proposed patch: iterating `source_event_ids` regresses redacted sources. Normalization prunes a redacted source's metadata entry by design, so requiring an entry for every `source_event_ids` member makes any coalesced turn with a redaction permanently unsuppressible, even though a redacted source owes no reply. The fix iterates `replay_source_event_ids` instead, which already drops redacted sources. Mutating it back to `source_event_ids` fails the new `redacted-False` and `redacted-True` matrix cases.

The rule now lives on `TurnRecord.requester_id_for_source`, shared with `edit_regenerator`, which had been enforcing the same per-source fail-closed rule inline. Sparse-metadata regression added as the `partial` case of `test_backlog_replay_respects_coalesced_source_ownership` (× `degraded`); mutating the miss branch to fall back to the turn-level requester fails both that case and `test_partial_coalesced_metadata_rejects_anchor_sender_editing_sibling`.
